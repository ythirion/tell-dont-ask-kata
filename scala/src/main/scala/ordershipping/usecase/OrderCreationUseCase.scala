package ordershipping.usecase

import ordershipping.domain.Order
import ordershipping.repository.{OrderRepository, ProductCatalog}

class OrderCreationUseCase(
                            val orderRepository: OrderRepository,
                            val productCatalog: ProductCatalog
                          ) {
  def run(request: SellItemsRequest): Unit = {
    // TODO : simplify it -> split
    val items = request.requests
      .map(request => {
        productCatalog.getByName(request.productName) match {
          case Some(p) => (p, request.quantity)
          case _ => throw new UnknownProductException
        }
      })
      .toMap

    val order = Order.create(items)

    orderRepository.save(order)
  }

  private def roundAt(p: Int)(n: Double): Double = {
    val s = math pow(10, p)
    (math round n * s) / s
  }
}
