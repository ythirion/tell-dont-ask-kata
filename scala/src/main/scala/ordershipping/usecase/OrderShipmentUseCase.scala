package ordershipping.usecase

import ordershipping.repository.OrderRepository
import ordershipping.service.ShipmentService

class OrderShipmentUseCase(
    val orderRepository: OrderRepository,
    val shipmentService: ShipmentService
) {
  def run(request: OrderShipmentRequest): Unit = {
    orderRepository
      .getById(request.orderId)
      .foreach(order => {
        order.ship(shipmentService)
        orderRepository.save(order)
      })
  }
}
