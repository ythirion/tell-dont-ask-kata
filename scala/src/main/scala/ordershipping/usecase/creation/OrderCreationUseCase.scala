package ordershipping.usecase.creation

import ordershipping.domain.{Order, Product}
import ordershipping.usecase.ports.repository.{OrderRepository, ProductCatalog}

class OrderCreationUseCase(
    val orderRepository: OrderRepository,
    val productCatalog: ProductCatalog
) {
  def run(request: SellItemsRequest): Unit = {
    orderRepository.save(
      Order.create(
        toProductQuantities(request)
      )
    )
  }

  private def toProductQuantities(
      request: SellItemsRequest
  ): Map[Product, Int] =
    request.requests
      .map(toProductQuantity)
      .toMap

  private def toProductQuantity(
      sellItemRequest: SellItemRequest
  ): (Product, Int) =
    productCatalog.getByName(sellItemRequest.productName) match {
      case Some(p) => (p, sellItemRequest.quantity)
      case _       => throw new UnknownProductException
    }
}
