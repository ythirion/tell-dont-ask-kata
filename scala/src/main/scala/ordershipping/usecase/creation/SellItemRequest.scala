package ordershipping.usecase.creation

case class SellItemRequest(productName: String, quantity: Int)
