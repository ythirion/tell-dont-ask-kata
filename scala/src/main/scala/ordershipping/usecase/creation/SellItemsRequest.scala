package ordershipping.usecase.creation

case class SellItemsRequest(requests: List[SellItemRequest])
