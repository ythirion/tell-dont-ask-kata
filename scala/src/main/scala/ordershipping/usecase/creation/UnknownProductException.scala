package ordershipping.usecase.creation

final case class UnknownProductException() extends Exception
