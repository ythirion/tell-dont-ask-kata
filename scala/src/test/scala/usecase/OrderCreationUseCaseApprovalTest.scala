package usecase

import builders.ProductBuilder._
import doubles.{InMemoryProductCatalog, TestOrderRepository}
import ordershipping.usecase.creation.{OrderCreationUseCase, SellItemRequest, SellItemsRequest}
import org.approvaltests.Approvals._
import org.junit.jupiter.api.{BeforeEach, Test}

class OrderCreationUseCaseApprovalTest {
  private val productCatalog = new InMemoryProductCatalog(
    List(
      food("salad").costing(3.56).build(),
      food("tomato").costing(4.65).build()
    )
  )
  private var orderRepository: TestOrderRepository = _
  private var useCase: OrderCreationUseCase = _

  @BeforeEach
  def setup(): Unit = {
    orderRepository = new TestOrderRepository()
    useCase = new OrderCreationUseCase(
      orderRepository = orderRepository,
      productCatalog = productCatalog
    )
  }

  @Test
  def OrderCreationUseCaseShouldSellMultipleItems(): Unit = {
    val saladRequest = SellItemRequest(productName = "salad", quantity = 2)
    val tomatoRequest = SellItemRequest(productName = "tomato", quantity = 3)

    val request = SellItemsRequest(
      List(saladRequest, tomatoRequest)
    )

    useCase.run(request)

    verify(
      orderRepository.savedOrder()
    )
  }
}
