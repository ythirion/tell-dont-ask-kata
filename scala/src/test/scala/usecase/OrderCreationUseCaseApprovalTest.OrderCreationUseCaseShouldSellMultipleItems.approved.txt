
Status:Created
Currency:EUR
Items:{Name:salad, Price:3.56, Quantity:2, Tax:0.72, TaxedAmount:7.84 ; Name:tomato, Price:4.65, Quantity:3, Tax:1.41, TaxedAmount:15.36}
Tax:2.13
Total:23.2
